File: src/main/java/puzzles/year23/day10/Direction.java
package puzzles.year23.day10;

public enum Direction {
    NORTH,
    SOUTH,
    EAST,
    WEST;

    public static Direction comingFrom(int x, int y, int prevX, int prevY) {
        int dX = x - prevX;
        int dY = y - prevY;

        if (dY == -1) {
            return SOUTH;
        } else if (dY == 1) {
            return NORTH;
        } else if (dX == -1) {
            return EAST;
        } else if (dX == 1) {
            return WEST;
        }

        throw new RuntimeException("Can't convert direction");
    }

    public static int[] nextPos(Direction direction, int x, int y) {
        switch (direction) {
            case NORTH -> {
                return new int[]{x, y - 1};
            }
            case SOUTH -> {
                return new int[]{x, y + 1};
            }
            case EAST -> {
                return new int[]{x + 1, y};
            }
            case WEST -> {
                return new int[]{x - 1, y};
            }
        }

        throw new RuntimeException("Shouldn't be able to get here");
    }
}

File: src/main/java/puzzles/year24/day6/Direction.kt
package puzzles.year24.day6

enum class Direction {
    NORTH,
    SOUTH,
    EAST,
    WEST;

    companion object {
        fun comingFrom(x: Int, y: Int, prevX: Int, prevY: Int): Direction {
            val dX = x - prevX
            val dY = y - prevY

            if (dY == -1) {
                return SOUTH
            } else if (dY == 1) {
                return NORTH
            } else if (dX == -1) {
                return EAST
            } else if (dX == 1) {
                return WEST
            }

            throw RuntimeException("Can't convert direction")
        }

        fun nextPos(direction: Direction?, pos: Pair<Int, Int>): Pair<Int, Int> {
            when (direction) {
                NORTH -> {
                    return Pair(pos.first, pos.second - 1)
                }

                SOUTH -> {
                    return Pair(pos.first, pos.second + 1)
                }

                EAST -> {
                    return Pair(pos.first + 1, pos.second)
                }

                WEST -> {
                    return Pair(pos.first - 1, pos.second)
                }

                null -> throw RuntimeException("Shouldn't be able to get here")
            }
        }

        fun nextDirection(direction: Direction): Direction {
            return when (direction) {
                NORTH -> EAST;
                EAST -> SOUTH;
                SOUTH -> WEST;
                WEST -> NORTH;
            }
        }
    }
}
